// ts-simple-ast/dist/compiler/tools/results/DocumentSpan.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var utils_1 = require("../../../utils");
var TextSpan_1 = require("./TextSpan");
/**
 * Document span.
 */
var DocumentSpan = /** @class */ (function () {
    /**
     * @internal
     */
    function DocumentSpan(global, compilerObject) {
        this.global = global;
        this._compilerObject = compilerObject;
        // store this node so that it's start doesn't go out of date because of manipulation (though the text span may)
        this.sourceFile = this.global.compilerFactory.getSourceFileFromCacheFromFilePath(this.compilerObject.fileName);
        this.node = this.getSourceFile().getDescendantAtStartWithWidth(this.getTextSpan().getStart(), this.getTextSpan().getLength());
    }
    Object.defineProperty(DocumentSpan.prototype, "compilerObject", {
        /**
         * Gets the compiler object.
         */
        get: function () {
            return this._compilerObject;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Gets the source file this reference is in.
     */
    DocumentSpan.prototype.getSourceFile = function () {
        return this.sourceFile;
    };
    /**
     * Gets the text span.
     */
    DocumentSpan.prototype.getTextSpan = function () {
        return new TextSpan_1.TextSpan(this.compilerObject.textSpan);
    };
    /**
     * Gets the node at the start of the text span.
     */
    DocumentSpan.prototype.getNode = function () {
        return this.node;
    };
    tslib_1.__decorate([
        utils_1.Memoize
    ], DocumentSpan.prototype, "getTextSpan", null);
    return DocumentSpan;
}());
exports.DocumentSpan = DocumentSpan;
