// ts-simple-ast/dist/utils/collections/Es5Map.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var Es5PropSaver_1 = require("../Es5PropSaver");
var Es5Map = /** @class */ (function () {
    function Es5Map() {
        this.propSaver = new Es5PropSaver_1.Es5PropSaver();
        this.items = {};
        this.itemCount = 0;
    }
    Object.defineProperty(Es5Map.prototype, "size", {
        get: function () {
            return Object.keys(this.items).length;
        },
        enumerable: true,
        configurable: true
    });
    Es5Map.prototype.set = function (key, value) {
        var identifier = this.getIdentifier(key) || this.createIdentifier(key);
        this.items[identifier] = [key, value];
    };
    Es5Map.prototype.get = function (key) {
        var identifier = this.getIdentifier(key);
        if (identifier == null)
            return undefined;
        var keyValue = this.items[identifier];
        if (keyValue == null)
            return undefined;
        return keyValue[1];
    };
    Es5Map.prototype.has = function (key) {
        var identifier = this.getIdentifier(key);
        if (identifier == null)
            return false;
        return this.items.hasOwnProperty(identifier);
    };
    Es5Map.prototype.delete = function (key) {
        var identifier = this.getIdentifier(key);
        if (identifier != null)
            delete this.items[identifier];
    };
    Es5Map.prototype.clear = function () {
        this.propSaver = new Es5PropSaver_1.Es5PropSaver();
        this.items = {};
    };
    Es5Map.prototype.entries = function () {
        var _a, _b, key, e_1_1, e_1, _c;
        return tslib_1.__generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    _d.trys.push([0, 5, 6, 7]);
                    _a = tslib_1.__values(Object.keys(this.items)), _b = _a.next();
                    _d.label = 1;
                case 1:
                    if (!!_b.done) return [3 /*break*/, 4];
                    key = _b.value;
                    return [4 /*yield*/, this.items[key]];
                case 2:
                    _d.sent();
                    _d.label = 3;
                case 3:
                    _b = _a.next();
                    return [3 /*break*/, 1];
                case 4: return [3 /*break*/, 7];
                case 5:
                    e_1_1 = _d.sent();
                    e_1 = { error: e_1_1 };
                    return [3 /*break*/, 7];
                case 6:
                    try {
                        if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
                    }
                    finally { if (e_1) throw e_1.error; }
                    return [7 /*endfinally*/];
                case 7: return [2 /*return*/];
            }
        });
    };
    Es5Map.prototype.keys = function () {
        var _a, _b, entry, e_2_1, e_2, _c;
        return tslib_1.__generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    _d.trys.push([0, 5, 6, 7]);
                    _a = tslib_1.__values(this.entries()), _b = _a.next();
                    _d.label = 1;
                case 1:
                    if (!!_b.done) return [3 /*break*/, 4];
                    entry = _b.value;
                    return [4 /*yield*/, entry[0]];
                case 2:
                    _d.sent();
                    _d.label = 3;
                case 3:
                    _b = _a.next();
                    return [3 /*break*/, 1];
                case 4: return [3 /*break*/, 7];
                case 5:
                    e_2_1 = _d.sent();
                    e_2 = { error: e_2_1 };
                    return [3 /*break*/, 7];
                case 6:
                    try {
                        if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
                    }
                    finally { if (e_2) throw e_2.error; }
                    return [7 /*endfinally*/];
                case 7: return [2 /*return*/];
            }
        });
    };
    Es5Map.prototype.values = function () {
        var _a, _b, entry, e_3_1, e_3, _c;
        return tslib_1.__generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    _d.trys.push([0, 5, 6, 7]);
                    _a = tslib_1.__values(this.entries()), _b = _a.next();
                    _d.label = 1;
                case 1:
                    if (!!_b.done) return [3 /*break*/, 4];
                    entry = _b.value;
                    return [4 /*yield*/, entry[1]];
                case 2:
                    _d.sent();
                    _d.label = 3;
                case 3:
                    _b = _a.next();
                    return [3 /*break*/, 1];
                case 4: return [3 /*break*/, 7];
                case 5:
                    e_3_1 = _d.sent();
                    e_3 = { error: e_3_1 };
                    return [3 /*break*/, 7];
                case 6:
                    try {
                        if (_b && !_b.done && (_c = _a.return)) _c.call(_a);
                    }
                    finally { if (e_3) throw e_3.error; }
                    return [7 /*endfinally*/];
                case 7: return [2 /*return*/];
            }
        });
    };
    Es5Map.prototype.getIdentifier = function (key) {
        if (typeof key === "string")
            return key;
        return this.propSaver.get(key);
    };
    Es5Map.prototype.createIdentifier = function (key) {
        if (typeof key === "string")
            return key;
        var identifier = (this.itemCount++).toString();
        this.propSaver.set(key, identifier);
        return identifier;
    };
    return Es5Map;
}());
exports.Es5Map = Es5Map;
