// @worldsibu/hurley/dist/generators/chaincodeinteractor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var invokechaincode_sh_1 = require("./invokechaincode.sh");
var ChaincodeInteractor = (function () {
    function ChaincodeInteractor(name, fn, options) {
        var args = [];
        for (var _i = 3; _i < arguments.length; _i++) {
            args[_i - 3] = arguments[_i];
        }
        this.name = name;
        this.fn = fn;
        this.options = options;
        this.invokeScript = new invokechaincode_sh_1.InvokeChaincodeShGenerator(options.networkRootPath, {
            channel: this.options.channel || 'ch1',
            name: name,
            function: fn,
            transientData: options.transientData,
            networkRootPath: options.networkRootPath,
            params: args || [],
            hyperledgerVersion: options.hyperledgerVersion,
            insideDocker: this.options.insideDocker,
            user: this.options.user, organization: this.options.organization
        });
    }
    ChaincodeInteractor.prototype.invoke = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                return [2, this.invokeScript.run()];
            });
        });
    };
    return ChaincodeInteractor;
}());
exports.ChaincodeInteractor = ChaincodeInteractor;
