// ts-simple-ast/dist/compiler/type/TypeParameterDeclaration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var typescript_1 = require("../../typescript");
var manipulation_1 = require("../../manipulation");
var base_1 = require("../base");
var common_1 = require("../common");
exports.TypeParameterDeclarationBase = base_1.NamedNode(common_1.Node);
var TypeParameterDeclaration = /** @class */ (function (_super) {
    tslib_1.__extends(TypeParameterDeclaration, _super);
    function TypeParameterDeclaration() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Gets the constraint node of the type parameter.
     */
    TypeParameterDeclaration.prototype.getConstraintNode = function () {
        return this.getNodeFromCompilerNodeIfExists(this.compilerNode.constraint);
    };
    /**
     * Gets the default node of the type parameter.
     */
    TypeParameterDeclaration.prototype.getDefaultNode = function () {
        return this.getNodeFromCompilerNodeIfExists(this.compilerNode.default);
    };
    /**
     * Removes this type parameter.
     */
    TypeParameterDeclaration.prototype.remove = function () {
        var parentSyntaxList = this.getParentSyntaxListOrThrow();
        var typeParameters = parentSyntaxList.getChildrenOfKind(typescript_1.SyntaxKind.TypeParameter);
        if (typeParameters.length === 1)
            removeAllTypeParameters();
        else
            manipulation_1.removeCommaSeparatedChild(this);
        function removeAllTypeParameters() {
            var children = [
                parentSyntaxList.getPreviousSiblingIfKindOrThrow(typescript_1.SyntaxKind.LessThanToken),
                parentSyntaxList,
                parentSyntaxList.getNextSiblingIfKindOrThrow(typescript_1.SyntaxKind.GreaterThanToken)
            ];
            manipulation_1.removeChildren({ children: children });
        }
    };
    return TypeParameterDeclaration;
}(exports.TypeParameterDeclarationBase));
exports.TypeParameterDeclaration = TypeParameterDeclaration;
